packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,236 @@
+    if (!this.l1TokenInfo[l1Token]) throw new Error("Token info not found. Ensure to construct correctly");
+    const { tokenType, tokenEthPrice } = this.l1TokenInfo[l1Token];
+
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Considering relay profitability",
+      l1Token,
+      tokenType: TokenType[tokenType],
+      tokenEthPrice: fromWei(tokenEthPrice),
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      slowRevenue: slowRevenue.toString(),
+      speedUpRevenue: speedUpRevenue.toString(),
+      instantRevenue: instantRevenue.toString(),
+    });
+
+    // Calculate the transaction submission costs depending on the gas price and the token type being relayed. This
+    // considers the relayer discount, applied to all relay actions equally.
+    const ethSubmissionCost = this.getRelayEthSubmissionCost(cumulativeGasPrice, tokenType);
+
+    // Calculate the expected revenue, in ETH, based on the amount of tokens being offered in fees.
+    const ethRevenue = this.getEthRevenue(tokenEthPrice, slowRevenue, speedUpRevenue, instantRevenue);
+
+    // Calculate the relay profitability as the difference between revenue and cost.
+    const ethProfitability = this.getProfit(ethSubmissionCost, ethRevenue);
+
+    // Finally, determine the relay submit type based on the profitability.
+    let relaySubmitType: RelaySubmitType;
+    if (
+      ethProfitability.instantEthProfit.gt(toBN(0)) &&
+      ethProfitability.instantEthProfit.gt(ethProfitability.slowEthProfit)
+    )
+      relaySubmitType = RelaySubmitType.Instant;
+    else if (ethProfitability.speedUpEthProfit.gt(toBN(0))) relaySubmitType = RelaySubmitType.SpeedUp;
+    else if (ethProfitability.slowEthProfit.gt(toBN(0))) relaySubmitType = RelaySubmitType.Slow;
+    else relaySubmitType = RelaySubmitType.Ignore;
+
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Considered relay profitability",
+      l1Token,
+      tokenType: TokenType[tokenType],
+      tokenEthPrice: fromWei(tokenEthPrice),
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      ethSubmissionCost: objectMap(ethSubmissionCost, (value: BN) => fromWei(value)),
+      ethRevenue: objectMap(ethRevenue, (value: BN) => fromWei(value)),
+      ethProfitability: objectMap(ethProfitability, (value: BN) => fromWei(value)),
+      relaySubmitType: RelaySubmitType[relaySubmitType],
+    });
+    return relaySubmitType;
+  }
+
+  getRelayEthSubmissionCost(
+    gasPrice: BN,
+    tokenType: TokenType
+  ): {
+    slowEThCost: BN;
+    speedUpEthCost: BN;
+    instantEthCost: BN;
+  } {
+    const discount = fixedPoint.sub(this.relayerDiscount);
+    switch (tokenType) {
+      case TokenType.WETH:
+        return {
+          slowEThCost: gasPrice.mul(toBN(across.constants.SLOW_ETH_GAS)).mul(discount).div(fixedPoint),
```

Comment:
Do you think this would be easier to read if we put all of these constants into a map, like:
```ts
const costConstants = {
  [TokenType.WETH]: {
    slowEthGas: across.constants.SLOW_ETH_GAS
    ...
```

Then down here, you could use your object mapper or some other mechanism to write the math once?

Thoughts? This just seems a little repetitive.

Reply:
sure. I did this + a bit more of a refactor to make it even fewer lines.